core/src/avm2/globals/toplevel.rs

```diff
@@ -379,3 +379,90 @@ fn encode_utf8_with_exclusions<'gc>(
 
     Ok(AvmString::new_utf8(activation.context.gc_context, output).into())
 }
+
+pub fn decode_uri<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    _this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    decode(
+        activation,
+        args,
+        // Characters that are reserved, sourced from as3 docs
+        "#$&+,/:;=?@",
+        "decodeURI",
+    )
+}
+
+pub fn decode_uri_component<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    _this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    decode(activation, args, "", "decodeURIComponent")
+}
+
+fn handle_percent<I>(chars: &mut I) -> Option<u8>
+where
+    I: Iterator<Item = char>,
+{
+    let high = chars.next()?.to_digit(16)? as u8;
+    let low = chars.next()?.to_digit(16)? as u8;
+    Some(low | (high << 4))
+}
+
+// code derived from flash.utils.unescapeMultiByte
+fn decode<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    args: &[Value<'gc>],
+    reserved_set: &str,
+    func_name: &str,
+) -> Result<Value<'gc>, Error<'gc>> {
+    let value = match args.first() {
+        None => return Ok("undefined".into()),
+        Some(Value::Undefined) => return Ok("null".into()),
+        Some(value) => value.coerce_to_string(activation)?,
+    };
+
+    let bs = value.as_wstr();
+    let mut output = String::new();
+    let chars = bs.chars().map(|c| c.unwrap_or(char::REPLACEMENT_CHARACTER));
+
+    let mut chars = chars.peekable();
+
+    while let Some(c) = chars.next() {
+        if c != '%' {
+            output.push(c);
+            continue;
+        }
+
+        let mut bytes = Vec::new();
+        while let Some(b) = handle_percent(&mut chars) {
+            bytes.push(b);
+            if !matches!(chars.peek(), Some('%')) {
+                break;
+            }
+            chars.next();
+        }
+        let decoded = String::from_utf8_lossy(&bytes);
```

Comment:
I believe you can do `std::str::from_utf8` here and return the `URIError` immediately on failure.